File: src/com/anathema_roguelike/main/ui/uielements/UIElement.java
/*******************************************************************************
 * Copyright (C) 2017 Carl Minden
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/
package com.anathema_roguelike.main.ui.uielements;

import java.util.ArrayList;

import com.anathema_roguelike.main.Game;
import com.anathema_roguelike.main.display.Color;
import com.anathema_roguelike.main.display.Display.DisplayLayer;
import com.anathema_roguelike.main.display.animations.Animation;
import com.anathema_roguelike.main.display.Outline;
import com.anathema_roguelike.main.display.RenderSurface;
import com.anathema_roguelike.main.display.Renderable;
import com.anathema_roguelike.main.utilities.position.HasPosition;
import com.anathema_roguelike.main.utilities.position.Point;

import squidpony.squidgrid.gui.gdx.SColor;

public abstract class UIElement extends RenderSurface implements Renderable, HasPosition {
	
	private Point position;
	private Point outerPosition;
	private int outerWidth;
	private int outerHeight;
	private float background;
	private Border border;
	
	private ArrayList<UIElement> elements = new ArrayList<>();
	
	public UIElement(Point position, int width, int height, float background) {
		super(width, height);
		
		init(position, position, width, height, background);
	}
	
	public UIElement(Point position, int width, int height, String borderTitle, float background) {
		super(width - 2, height - ((borderTitle == null) ? 2 : 4));
		
		init(new Point(position.getX() + 1, position.getY() + ((borderTitle == null) ? 1 : 3)), position, width, height, background);
		
		this.border = new Border(this, borderTitle);
	}
	
	private void init(Point position, Point outerPosition, int width, int height, float background) {
		this.position = position;
		this.outerPosition = outerPosition;
		this.outerWidth = width;
		this.outerHeight = height;
		this.background = background;
	}
	
	protected abstract void renderContent();
	
	protected void renderBackground() {
		
		if(border != null) {
			border.render();
		}
		
		for(int i = 0; i < getWidth(); i++) {
			for(int j = 0; j < getHeight(); j++) {
				
				if(background == 0f) {
					renderChar(DisplayLayer.UI_BACKGROUND, i, j, ' ', Color.BLACK);
				} else if(background == 1f) {
					renderChar(DisplayLayer.UI_BACKGROUND, i, j, '\u2588', Color.BLACK);
				} else {
					renderChar(DisplayLayer.UI_BACKGROUND, i, j, '\u2588', Color.opacity(Color.BLACK, background));
				}
			}
		}
	}
	
	@Override
	public Point getPosition() {
		return position;
	}
		
	public void addUIElement(UIElement element) {
		elements.add(element);
	}
	
	public void removeUIElement(UIElement element) {
		elements.remove(element);
	}
	
	public ArrayList<UIElement> getUIElements() {
		return elements;
	}
	
	public int getOuterX() {
		return outerPosition.getX();
	}

	public int getOuterY() {
		return outerPosition.getY();
	}

	public int getOuterWidth() {
		return outerWidth;
	}

	public int getOuterHeight() {
		return outerHeight;
	}
	
	public float getBackground() {
		return background;
	}

	@Override
	public final void render() {
		
		renderBackground();
		renderContent();
		
		for(UIElement element : new ArrayList<>(getUIElements())) {
			element.render();
		}
		
		//occasionally useful UI debugging tool
		
//		renderChar(DisplayLayer.UI_FOREGROUND, 0, 0, 'X', Color.ERROR);
//		renderChar(DisplayLayer.UI_FOREGROUND, 0, getHeight() - 1, 'X', Color.ERROR);
//		renderChar(DisplayLayer.UI_FOREGROUND, getWidth() - 1, 0, 'X', Color.ERROR);
//		renderChar(DisplayLayer.UI_FOREGROUND, getWidth() - 1, getHeight() - 1, 'X', Color.ERROR);
	}
	
	@Override
	public void renderChar(DisplayLayer layer, int x, int y, char string, SColor color) {
		Game.getInstance().getDisplay().renderChar(layer, x + getX(), y + getY(), string, color);
	}
	
	@Override
	public void createAnimation(DisplayLayer layer, Animation animation) {
		super.createAnimation(layer, animation, getPosition());
	}
	
	public void renderOutline(Outline outline) {
		outline.setOffset(new Point(getX(), getY()));
		outline.render();
	}
}

File: src/com/anathema_roguelike/main/ui/uielements/UIElement.scala
/*******************************************************************************
 * Copyright (c) 2019. Carl Minden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package com.anathema_roguelike
package main.ui.uielements

import java.util
import java.util.ArrayList

import com.anathema_roguelike.main.Game
import com.anathema_roguelike.main.display.Display.DisplayLayer
import com.anathema_roguelike.main.display.{Color, Display, Outline, RenderSurface, Renderable}
import com.anathema_roguelike.main.display.animations.Animation
import com.anathema_roguelike.main.utilities.position.{HasPosition, Point}
import squidpony.squidgrid.gui.gdx.SColor

import scala.collection.mutable.ListBuffer

abstract class UIElement private (
      position: Point,
      outerPosition: Point,
      width: Int,
      outerWidth: Int,
      height: Int,
      outerHeight: Int,
      background: Float
  ) extends RenderSurface(width, height) with Renderable with HasPosition {

  private val elements: ListBuffer[UIElement] = ListBuffer[UIElement]()
  private var border: Option[Border] = None

  def this(position: Point, width: Int, height: Int, background: Float) {
    this(position, position, width, width, height, height, background)
  }

  def this(position: Point, width: Int, height: Int, background: Float, borderTitle: Option[String]) {
    this(
      Point(position.getX + 1, position.getY + borderTitle.map(_ => 3).getOrElse(1)),
      position,
      width - 2,
      width,
      height - borderTitle.map(_ => 4).getOrElse(2),
      height,
      background
    )

    border = borderTitle.map(b => new Border(this, b))
  }

  protected def renderContent(): Unit

  protected def renderBackground(): Unit = {

    border.foreach(_.render())

    for (i <- 0 until getWidth; j <- 0 until getHeight) {
      if (background == 0f) renderChar(DisplayLayer.UI_BACKGROUND, i, j, ' ', Color.BLACK)
      else if (background == 1f) renderChar(DisplayLayer.UI_BACKGROUND, i, j, '\u2588', Color.BLACK)
      else renderChar(DisplayLayer.UI_BACKGROUND, i, j, '\u2588', Color.opacity(Color.BLACK, background))
    }
  }

  override def getPosition: Point = position

  def addUIElement(element: UIElement): Unit = {
    elements += element
  }

  def removeUIElement(element: UIElement): Unit = {
    elements -= element
  }

  def getUIElements: Iterable[UIElement] = elements

  def getOuterX: Int = outerPosition.getX

  def getOuterY: Int = outerPosition.getY

  def getOuterWidth: Int = outerWidth

  def getOuterHeight: Int = outerHeight

  def getBackground: Float = background

  override final def render(): Unit = {
    renderBackground()
    renderContent()

    for (element <- getUIElements) {
      element.render()
    }
    //occasionally useful UI debugging tool
    //		renderChar(DisplayLayer.UI_FOREGROUND, 0, 0, 'X', Color.ERROR);
    //		renderChar(DisplayLayer.UI_FOREGROUND, 0, getHeight() - 1, 'X', Color.ERROR);
    //		renderChar(DisplayLayer.UI_FOREGROUND, getWidth() - 1, 0, 'X', Color.ERROR);
    //		renderChar(DisplayLayer.UI_FOREGROUND, getWidth() - 1, getHeight() - 1, 'X', Color.ERROR);
  }

  override def renderChar(layer: Display.DisplayLayer, x: Int, y: Int, string: Char, color: SColor): Unit = {
    Game.getInstance.getDisplay.renderChar(layer, x + getX, y + getY, string, color)
  }

  def createAnimation(layer: Display.DisplayLayer, animation: Animation): Unit = {
    super.createAnimation(layer, animation, getPosition)
  }

  override def renderOutline(outline: Outline): Unit = {
    outline.setOffset(Point(getX, getY))
    outline.render()
  }
}
